packages/picasso.js/src/core/chart-components/ref-line/lines-and-labels.js

```diff
@@ -19,6 +19,95 @@ export function refLabelDefaultSettings() {
   };
 }
 
+function isMaxY(chart, slope, value) {
+  // when maxY exceeds the shown scale
+  const scaleX = chart.scale('x');
+  const scaleY = chart.scale('y');
+  if (slope > 0) {
+    const maxY = scaleX.max() * slope + value;
+    if (maxY >= scaleY.max()) {
+      return true;
+    }
+  } else if (slope < 0) {
+    const minY = scaleX.max() * slope + value;
+    if (minY <= scaleY.min()) {
+      return true;
+    }
+  }
+  return false;
+}
+function getMaxXPosition(chart, slope, value) {
+  // if maxY then get maxX position available on the scale
+  const scaleX = chart.scale('x');
+  const scaleY = chart.scale('y');
+  // For negative slopes
+  if (slope < 0) {
+    return scaleX((scaleY.min() - value) / slope);
+  }
+  // For positive slopes
+  return scaleX((scaleY.max() - value) / slope);
+}
+
+function getFormatter(p, chart) {
+  let formatter;
+  if (typeof p.formatter === 'string') {
+    formatter = chart.formatter(p.formatter);
+  } else if (typeof p.formatter === 'object') {
+    formatter = chart.formatter(p.formatter);
+  } else if (typeof p.scale !== 'undefined' && p.scale.data) {
+    // TODO - Add support for array as source into formatter
+    const scaleData = p.scale.data() && p.scale.data().fields;
+    formatter = scaleData && scaleData[0] ? scaleData[0].formatter() : null;
+  }
+  return formatter;
+}
+
+function isColliding(items, slopeValue, slope, measured, maxX, xPadding, yPadding) {
+  for (let i = 0, len = items.length; i < len; i++) {
+    let curItem = items[i];
+    if (curItem?.type === 'text') {
+      if (slope > 0 && maxX !== undefined) {
+        if (
+          Math.abs(curItem.x - slopeValue.x) < curItem.width + xPadding &&
+          Math.abs(curItem.y - slopeValue.y) < measured.height + yPadding
+        ) {
+          return true;
+        }
+      } else if (Math.abs(curItem.y - slopeValue.y) < measured.height + yPadding) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+function calculateX(slopeLine, slope, maxX, measured, blueprint, slopeStyle) {
+  // calculate x for the various scenarios possible
+  let x;
```

Comment:
Looks like you can `return` instead of reassigning `x`, that should reduce the complixity a little bit